File: C++/Question4.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

int findLongCommPreLen(const int* firstArray, int firstArraySize, const int* secondArray, int secondArraySize) {
    int longestCommonPrefixLength = 0;

    for (int i = 0; i < firstArraySize; i++) {
        for (int j = 0; j < secondArraySize; j++) {
            string firstNumStr = to_string(firstArray[i]);
            string secondNumStr = to_string(secondArray[j]);

            int prefixLength = 0;
            int index = 0;

            while (index < firstNumStr.length() && index < secondNumStr.length() && firstNumStr[index] == secondNumStr[index]) {
                prefixLength++;
                index++;
            }

            longestCommonPrefixLength = max(longestCommonPrefixLength, prefixLength);
        }
    }

    return longestCommonPrefixLength;
}

int* parseInput(const string& input, int& size) {
    istringstream iss(input);
    string numStr;

    size = 0;
    while (getline(iss, numStr, ',')) {
        size++;
    }

    int* nums = new int[size];
    iss.clear();
    iss.seekg(0);

    for (int i = 0; i < size; i++) {
        getline(iss, numStr, ',');
        nums[i] = stoi(numStr);
    }

    return nums;
}

int main() {
    string firstArrayInput;
    string secondArrayInput;

    cout << "";
    getline(cin, firstArrayInput);

    cout << "secondArray: ";
    getline(cin, secondArrayInput);

    int firstArraySize = 0;
    int* firstArray = parseInput(firstArrayInput, firstArraySize);

    int secondArraySize = 0;
    int* secondArray = parseInput(secondArrayInput, secondArraySize);

    int longestCommonPrefixLength = findLongCommPreLen(firstArray, firstArraySize, secondArray, secondArraySize);

    cout << "solution(firstArray, secondArray) = " << longestCommonPrefixLength << endl;

    delete[] firstArray;
    delete[] secondArray;

    return 0;
}

File: C++/Question4.py
def findLongCommPreLen(first_array, second_array):
    longest_common_prefix_length = 0

    for first_num in first_array:
        for second_num in second_array:
            first_num_str = str(first_num)
            second_num_str = str(second_num)

            prefix_length = 0
            index = 0

            while (index < len(first_num_str) and index < len(second_num_str) and first_num_str[index] == second_num_str[index]):
                prefix_length += 1
                index += 1

            longest_common_prefix_length = max(longest_common_prefix_length, prefix_length)

    return longest_common_prefix_length

def parse_input(input_string):
    nums = input_string.strip().split(", ")
    return [int(num) for num in nums]

def main():
    first_array_input = input("firstArray: ")
    second_array_input = input("secondArray: ")

    first_array = parse_input(first_array_input)
    second_array = parse_input(second_array_input)

    longest_common_prefix_length = findLongCommPreLen(first_array, second_array)

    print(f"solution(firstArray, secondArray) = {longest_common_prefix_length}")

if __name__ == "__main__":
    main()
